core/iwasm/interpreter/wasm_loader.c

```diff
@@ -250,36 +250,65 @@ loader_malloc(uint64 size, char *error_buf, uint32 error_buf_size)
 static bool
 check_utf8_str(const uint8* str, uint32 len)
 {
-    const uint8 *p = str, *p_end = str + len, *p_end1;
-    uint8 chr, n_bytes;
+    const uint8 *p = str, *p_end = str + len;
 
     while (p < p_end) {
-        chr = *p++;
-        if (chr >= 0x80) {
-            /* Calculate the byte count: the first byte must be
-               110XXXXX, 1110XXXX, 11110XXX, 111110XX, or 1111110X,
-               the count of leading '1' denotes the total byte count */
-            n_bytes = 0;
-            while ((chr & 0x80) != 0) {
-                chr = (uint8)(chr << 1);
-                n_bytes++;
-            }
-
-            /* Check byte count */
-            if (n_bytes < 2 || n_bytes > 6
-                || p + n_bytes - 1 > p_end)
+        if (*p < 0x80) {
+            p++;
+        }
+        else if (*p >= 0xC2 && *p <= 0xDF) {
+            if (p + 1 >= p_end || p[1] < 0x80 || p[1] > 0xBF) {
                 return false;
-
-            /* Check the following bytes, which must be 10XXXXXX */
-            p_end1 = p + n_bytes - 1;
-            while (p < p_end1) {
-                if (!(*p & 0x80) || (*p | 0x40))
+            }
+            p += 2;
+        }
+        else if (*p >= 0xE0 && *p <= 0xEF) {
+            if (*p == 0xE0) {
+                if (p + 2 >= p_end || p[1] < 0xA0 || p[1] > 0xBF || p[2] < 0x80
```

Comment:
p + 2 >= p_end is checked in every branch, could we move it to the beginning, after line 265?

Had better put p[2] < 0x80 to next line (or all in this line):
  if (p[1] < 0xA0 || p[1] > 0xBF
      || p[2]...)